proto/cmp/services/notification/v2/notify.proto

```diff
@@ -0,0 +1,119 @@
+//
+/// @custom:cmp-service type:product routing:local on-chain:true
+service NotificationService {
+  rpc TokenBoughtNotification(TokenBought) returns (google.protobuf.Empty);
+  rpc TokenExpiredNotification(TokenExpired) returns (google.protobuf.Empty);
+  rpc CancellationPendingNotification(CancellationPending) returns (google.protobuf.Empty);
+  rpc CancellationAcceptedNotification(CancellationAccepted) returns (google.protobuf.Empty);
+  rpc CancellationAcceptedByTheOwnerNotification(CancellationAcceptedByTheOwner) returns (google.protobuf.Empty);
+  rpc CancellationCounteredNotification(CancellationCountered) returns (google.protobuf.Empty);
+  rpc CancellationWithdrawnNotification(CancellationWithdrawn) returns (google.protobuf.Empty);
+  rpc CancellationRejectedNotification(CancellationRejected) returns (google.protobuf.Empty);
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event TokenBought(uint256 indexed tokenId, address indexed buyer);
+// ```
+message TokenBought {
+  uint64 token_id = 1;
+  string tx_id = 2;
+  cmp.types.v1.UUID mint_id = 3;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event TokenExpired(uint256 indexed tokenId);
+// ```
+message TokenExpired {
+  uint64 token_id = 1;
+  cmp.types.v1.UUID mint_id = 2;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event CancellationPending(uint256 indexed tokenId, address indexed proposedBy, uint256 refundAmount, uint16 cancellationReason, uint16 cancellationReasonVersion);
+// ```
+message CancellationPending {
+  uint64 token_id = 1;
+  string proposed_by = 2;
+  uint64 refund_amount = 3;
+  cmp.services.cancellation.v1.CancellationReason reason = 4;
+  string tx_id = 5;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event CancellationAccepted(uint256 indexed tokenId, address indexed acceptedBy, uint256 refundAmount);
+// ```
+message CancellationAccepted {
+  uint64 token_id = 1;
+  string accepted_by = 2;
+  uint64 refund_amount = 3;
+  string tx_id = 4;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event CancellationAcceptedByTheOwner(uint256 indexed tokenId, address indexed acceptedBy, uint256 refundAmount);
+// ```
+message CancellationAcceptedByTheOwner {
+  uint64 token_id = 1;
+  string accepted_by = 2;
+  uint64 refund_amount = 3;
+  string tx_id = 4;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event CancellationCountered(uint256 indexed tokenId, address indexed counteredBy, uint256 newRefundAmount);
+// ```
+message CancellationCountered {
+  uint64 token_id = 1;
+  string countered_by = 2;
+  uint64 new_refund_amount = 3;
+  string tx_id = 4;
+}
+
+// Related on-chain event:
+//
+// ```solidity
+// event CancellationWithdrawn(uint256 indexed tokenId, address indexed cancelledBy, uint16 reason, uint16 reasonVersion);
+// ```
+message CancellationWithdrawn {
+  uint64 token_id = 1;
+  string cancelled_by = 2;
+  cmp.services.cancellation.v1.WithdrawalReason reason = 3;
+  string tx_id = 4;
+}
+
```

Comment:
Same as the other comments. We don't need the reason in the protobuf messages because the version is already in the reason type used in the field above it. It's only needed on-chain because smart contract doesn't know about off-chain protobuf definitions.